### ISSUE TITLE
[template-functions]: add support for numeric operations

### ISSUE THREAD

**Issue by grrtrr** [go-easygen/easygen#9]
This is a feature request, for discussion and comments.

###  Use case

I ran into this when generating `enum` definitions for [protobuf](https://github.com/gogo/protobuf). Unlike go/c enums, they require to print out the numeric values, like:
```
enum OsCategory {
  LINUX   = 0;
  WINDOWS = 1;
  OTHER   = 2;
}
```

This requires more heavy lifting inside the `.tmpl` file - keeping track of start index, and adding sucessive index values to it.

Currently only `minus1` is defined. It would generally be more useful to have other arithmetic operations defined.

### Suggested improvement

Rather than re-inventing from scratch, I am suggesting to re-use e.g.
- [consul template functions](https://github.com/hashicorp/consul-template/blob/de2ebf4/template_functions.go#L727-L901) and/or
- import an existing package of template functions, e.g. [gtf](https://github.com/leekchan/gtf), which looks really useful.

**Comment by suntong** [comment]
Yes, this need some serious thinking. 

- For the solution specific to the enum definitions problem here, I think a closure would be sufficient (will provide a sample later). 
- To go any further, the consul template functions looks very good, and the gtf might also be as well. Just that I'm hesitating to introduce more dependencies to easygen. 
  * I had done some thinking along this path, and had once thought to refactor the `github.com/danverbraganza/varcaser/varcaser` dependency out to a sub module. I.e., for people to use it, they need to import `github.com/go-easygen/easygen` ***and*** `github.com/go-easygen/easygen/ext` or something like that. 
  * The ultimate goal is to make it simple for `easygen` to be a building block that people can easily extend, but I haven't figured out a good way of doing that yet. 
  * As for `minus1`, I had once thought to change it to just a generic `add` , which takes an extra parameter of `-1`, so _add -1_ becomes `minus1`. However, due to the limitation of go template, which accepts only one parameter, I was stopped. Maybe I didn't look far enough. So do you think that if we do something like [consul template functions](https://github.com/hashicorp/consul-template/blob/de2ebf4/template_functions.go#L727-L901), we can use those generic template functions?

It's a good discussion, I don't have the answer, :-)

**Comment by grrtrr** [comment]
Thank you for your comments. 

With regard to dependencies, I actually was thinking of adding the functions into `easygen`, not via dependencies. According to one of the go [proverbs](https://go-proverbs.github.io/), which says "a little copying is better than a little dependency".  The arithmetic functions in consul-template are great, they can be copied/integrated - as long if you are ok with the Mozilla Public License 2.0 (which admits copying, but requires to add the license to the source file). They could sit in a file like e.g. `tf-numeric.go`?

**Comment by suntong** [comment]
Ha, I really like the idea "_a little copying is better than a little dependency_", and I'm fine with the Mozilla Public License as well. 

Let me clarify this first, do you have something more in mind to tackle with, or only this specific enum definitions problem that we are discussing -- I do agree that consul-template are great, and we can copy them over. just that, is it necessary now, :-)

Also, have you looked into how to get round the go template accepting only one parameter barrier yet? Or is it just a misbelieve?

**Comment by suntong** [comment]
Here is an example of using Closures, 

https://github.com/suntong/lang/blob/master/lang/Go/src/ds/Closures.go#L36-L42

Which should solve this specific enum definitions problem perfectly.

**Comment by grrtrr** [comment]
Thank you for the suggestions. I worked around this by _requiring_ that all enums start at 0 and then continue with 1, 2, .... 

This makes a simpler definition possible using `{{ range $i, $j :=  .Values }}`, i.e. the value is just `$i`.
Arithmetic would only be needed if any of the enums has an _offset_ value, e.g. `[0, 200, 201, 500, 501]`.

Long story short, there is no pressing need for numeric functions now, it would still be a "nice to have" :-)

**Comment by suntong** [comment]
Great to hear. Glad you found an easier route. I mean, for the "nice to have", sure, we will get it when we need it, :-)

**Comment by suntong** [comment]
That consul template functions, when doing `{{subtract 9 2}}` will give `-7` instead of 7. 
Do you know why?

To get `7` I have to use `{{subtract 2 9}}`, that looks really weird to me, and I really really want to reverse the order. What do you think?

**Comment by suntong** [comment]
I'm almost there:
https://github.com/go-easygen/easygen/blob/restructure/example_test.go#L59-L62

Next it to wrap this up into standard APIs... 
https://github.com/go-easygen/easygen/blob/restructure/example_test.go#L61-L62

**Comment by suntong** [comment]
fixed in v2.1.0
